PhysicsTools/PatAlgos/plugins/PatMuonMerger.cc

```diff
@@ -0,0 +1,116 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+#include "DataFormats/Common/interface/getRef.h"
+#include "DataFormats/Common/interface/RefToPtr.h"
+
+namespace pat {
+  typedef edm::Ptr<pat::PackedCandidate> PackedCandidatePtr;
+}
+
+class PatMuonMerger : public edm::stream::EDProducer<> {
+public:
+  explicit PatMuonMerger(const edm::ParameterSet& iConfig);
+  ~PatMuonMerger() override {}
+
+  void produce(edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+
+private:
+  edm::InputTag muons_;
+  StringCutObjectSelector<pat::Muon, false> muonsCut_;
+  edm::InputTag pfCandidate_;
+  StringCutObjectSelector<pat::PackedCandidate, false> pfCandidateCut_;
+  edm::InputTag lostTrack_;
+  StringCutObjectSelector<pat::PackedCandidate, false> lostTrackCut_;
+
+  edm::EDGetTokenT<std::vector<pat::Muon>> muonToken_;
+  edm::EDGetTokenT<std::vector<pat::PackedCandidate>> pfCandToken_;
+  edm::EDGetTokenT<std::vector<pat::PackedCandidate>> lostTrackToken_;
+};
+
+PatMuonMerger::PatMuonMerger(const edm::ParameterSet& iConfig)
+    : muons_(iConfig.getParameter<edm::InputTag>("muons")),
+      muonsCut_(iConfig.existsAs<std::string>("muonCut") ? iConfig.getParameter<std::string>("muonCut") : ""),
```

Comment:
I think that a fillDescription() method should be eventually added also for the previous PATLostTracks producer. Since that producer already didn't have it, that could be deferred to a future PR.
However, this PatMuonMerger is a new class first implemented in this PR. Therefore, please add a fillDescriptions() here. You can get rid of all these "existsAs" afterwards.